### PR TITLE
Add rust-clippy to CI and set docker UID

### PR DESCRIPTION
**Pull request by anghelcovici**
Clippy compiles code and this took a long time since we were not reusing
artefacts. In order to do that, we are running everything in the
container as current user instead of root. This requires mounting both
the bazel cache and the cargo one.

Fixes: #70, #79

---
This change is [Reviewable](https://reviewable.io/reviews/project-oak/oak/86)

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -1 +1 @@
-/bazel-cache
+/*
```
Comment (tiziano88) [inline .dockerignore]: Maybe add a comment saying that we don't need anything from the docker context when building the image.

File: Dockerfile [hunk 2]
```diff
@@ -6,22 +6,34 @@ RUN git --version
 RUN clang-format -version
 RUN shellcheck --version
 
+# Create an user so we don't run this as root.
+ARG USERID
```
Comment (tiziano88) [inline Dockerfile]: (optional) UID, GID and USERNAME

File: Dockerfile [hunk 3]
```diff
@@ -6,22 +6,34 @@ RUN git --version
...
+ARG USERNAME
+ARG USERGRP
+
+RUN groupadd -g $USERGRP primarygroup
```
Comment (tiziano88) [inline Dockerfile]: Please use long options when possible, e.g. `--gid=$USERGRP` etc.

Comment (tiziano88) [inline Dockerfile]: If we are hardcoding `primarygroup` here we may as well hardcode the username, e.g. we may have `docker:docker` or something like that; in practice I don't think we rely on the username being consistent anyways.

File: scripts/docker_run [hunk 4]
```diff
@@ -3,11 +3,19 @@
 set -o errexit
 set -o xtrace
 
-docker build --tag=oak .
+USERNAME=$(id -un)
```
Comment (tiziano88) [inline scripts/docker_run]: Make `readonly`, and also quote RHS.

File: scripts/docker_run [hunk 5]
```diff
@@ -3,11 +3,19 @@
...
+
+mkdir -p $PWD/bazel-cache
```
Comment (tiziano88) [inline scripts/docker_run]: I think we already use `.` instead of `$PWD` in other places, so we should stick to that?

File: scripts/docker_run [hunk 6]
```diff
@@ -3,11 +3,19 @@
...
+touch $PWD/bazel-cache/.placeholder
```
Comment (tiziano88) [inline scripts/docker_run]: We either need to `mkdir`, *or* add the placeholder file and check it in, not both (I think).

File: scripts/docker_run [hunk 7]
```diff
@@ -3,11 +3,19 @@
...
+mkdir -p $PWD/cargo-cache
+touch $PWD/cargo-cache/.placeholder
+
+docker build --tag=oak --build-arg USERNAME="$USERNAME" --build-arg USERID="$(id -u)" --build-arg USERGRP="$(id -g)" .
```
Comment (tiziano88) [inline scripts/docker_run]: `--build-arg="USERNAME=$USERNAME"` (and below), and also one flag per line.

File: scripts/docker_run (hunk 4 above)
Comment (anghelcovici) [reply scripts/docker_run]: Done

File: .dockerignore (hunk 1 above)
Comment (anghelcovici) [reply .dockerignore]: Done

File: Dockerfile (hunk 2 above)
Comment (anghelcovici) [reply Dockerfile]: Done

File: Dockerfile (hunk 3 above)
Comment (anghelcovici) [reply Dockerfile]: Changed all

Comment (anghelcovici) [reply Dockerfile]: Done

File: scripts/docker_run (hunk 5 above)
Comment (anghelcovici) [reply scripts/docker_run]: We need to keep $PWD for docker volumes, as it doesn't work with .

File: scripts/docker_run (hunk 6 above)
Comment (anghelcovici) [reply scripts/docker_run]: Done

File: scripts/docker_run (hunk 7 above)
Comment (anghelcovici) [reply scripts/docker_run]: Done

File: scripts/docker_run [hunk 8]
```diff
@@ -3,11 +3,22 @@
 set -o errexit
 set -o xtrace
 
-docker build --tag=oak .
+readonly USERNAME="$(id --user --name)"
+
+mkdir -p "./bazel-cache"
```
Comment (tiziano88) [inline scripts/docker_run]: (nit) Single quotes around string literal.

File: scripts/docker_run [hunk 9]
```diff
@@ -3,11 +3,22 @@
...
+mkdir -p "./cargo-cache"
+
+docker build \
+  --tag=oak \
+  --build-arg "USERNAME=$USERNAME" \
```
Comment (tiziano88) [inline scripts/docker_run]: (nit) I would also add the `=` after `--build-arg` for consistency with all the other key-value flags.

File: scripts/docker_run (hunk 8 above)
Comment (anghelcovici) [reply scripts/docker_run]: Done

**Comment by anghelcovici** [comment]
Nice, so the travis integration failed now. I'll force commit, rebase should fix it.

**Pull request merged**